sdk/sync/client.go

```diff
@@ -0,0 +1,228 @@
+	DialTimeout:        10 * time.Second,
+	ReadTimeout:        10 * time.Second,
+	WriteTimeout:       10 * time.Second,
+	IdleCheckFrequency: 30 * time.Second,
+	MaxConnAge:         1 * time.Minute,
+}
+
+type Client struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+	wg     sync.WaitGroup
+
+	log       *zap.SugaredLogger
+	extractor func(ctx context.Context) (rp *runtime.RunParams)
+
+	rclient *redis.Client
+
+	barrierCh chan *newBarrier
+	newSubCh  chan *newSubscription
+}
+
+// NewBoundClient returns a new sync Client that is bound to the provided
+// RunEnv. All operations will be automatically scoped to the keyspace of that
+// run.
+//
+// The context passed in here will govern the lifecycle of the client.
+// Cancelling it will cancel all ongoing operations. However, for a clean
+// closure, the user should call Close().
+//
+// For test plans, a suitable context to pass here is the background context.
+func NewBoundClient(ctx context.Context, runenv *runtime.RunEnv) (*Client, error) {
+	return newClient(ctx, runenv.SLogger(), func(ctx context.Context) *runtime.RunParams {
+		return &runenv.RunParams
+	})
+}
+
+// MustBoundClient creates a new bound client by calling NewBoundClient, and
+// panicking if it errors.
+func MustBoundClient(ctx context.Context, runenv *runtime.RunEnv) *Client {
+	c, err := NewBoundClient(ctx, runenv)
+	if err != nil {
+		panic(err)
+	}
+	return c
+}
+
+// NewGenericClient returns a new sync Client that is bound to no RunEnv.
+// It is intended to be used by testground services like the sidecar.
+//
+// All operations expect to find the RunParams of the run to scope its actions
+// inside the supplied context.Context. Call WithRunParams to bind the
+// appropriate RunParams.
+//
+// The context passed in here will govern the lifecycle of the client.
+// Cancelling it will cancel all ongoing operations. However, for a clean
+// closure, the user should call Close().
+//
+// A suitable context to pass here is the background context of the main
+// process.
+func NewGenericClient(ctx context.Context, log *zap.SugaredLogger) (*Client, error) {
+	return newClient(ctx, log, GetRunParams)
+}
+
+// MustGenericClient creates a new generic client by calling NewGenericClient,
+// and panicking if it errors.
+func MustGenericClient(ctx context.Context, log *zap.SugaredLogger) *Client {
+	c, err := NewGenericClient(ctx, log)
+	if err != nil {
+		panic(err)
+	}
+	return c
+}
+
+// newClient creates a new sync client.
+func newClient(ctx context.Context, log *zap.SugaredLogger, extractor func(ctx context.Context) *runtime.RunParams) (*Client, error) {
+	rclient, err := redisClient(ctx, log)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create redis client: %w", err)
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	c := &Client{
+		ctx:       ctx,
+		cancel:    cancel,
+		log:       log,
+		extractor: extractor,
+		rclient:   rclient,
+		barrierCh: make(chan *newBarrier, 16),
```

Comment:
Why are `barrierCh` and `newSubCh` sizes set to 16? Generally I expect to see these channels unbuffered, so that delivery is guaranteed.

Reply:
Yeah, this might be a leftover from a previous design. Now that we make the caller wait anyway, these channels can be non-buffered.